cpp/src/arrow/csv/parser.cc

```diff
@@ -35,14 +35,13 @@ using detail::ParsedValueDesc;
 
 namespace {
 
-Status ParseError(const char* message) {
-  return Status::Invalid("CSV parse error: ", message);
+template <typename... Args>
+Status ParseError(Args&&... args) {
+  return Status::Invalid("CSV parse error: ", std::forward<Args>(args)...);
 }
 
-Status MismatchingColumns(int32_t expected, int32_t actual) {
-  char s[50];
-  snprintf(s, sizeof(s), "Expected %d columns, got %d", expected, actual);
-  return ParseError(s);
+Status MismatchingColumns(int32_t expected, int32_t actual, const std::string& row) {
```

Comment:
The row could be very long.  If we want to provide more context is there anyway we can put the row index in here instead?  Given that we only allow \n or \r\n to be line breaks that should make it pretty easy to locate the offending row via row index (i.e. row index == line number).

Reply:
Yes row index or line number which are not always the same because in theory new lines can be in a value would be nice but there didn't seem to be anything which tracked overall parser location at this level to be able to do that. That would also be good for the convert layer so that if a conversion error does occur not just column number but line/row number could be given too.

Reply:
Another option could be to just truncate `row` if it is too long.  Adding location tracking to put line numbers in errors could be a standalone JIRA / future PR as well.

Reply:
I'll add a truncate with an ellipse if over say 50 characters and open a jira ticket for tracking absolute line and row during csv parsing to be included in error messages